sw/host/opentitanlib/src/test_utils/bitbanging.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod i2c {
+    use anyhow::{anyhow, bail, Result};
+    use std::iter::Peekable;
+
+    #[derive(Debug, PartialEq)]
+    pub enum Transfer<'w> {
+        Start,
+        Stop,
+        Addr(u8, bool),
+        Write(&'w [u8]),
+        WriteNibble(u8), // Can be used to simulate a broken byte.
+        Read(usize),
+        Byte(u8, bool),
+        Bytes(&'w [u8]),
+    }
+
+    impl Transfer<'_> {
+        fn parse<const SDA: u8, const SCL: u8>(&self, is_next_stop: bool) -> Vec<u8> {
+            match *self {
+                Transfer::Start => vec![
+                    0x01 << SDA | 0x01 << SCL,
+                    0x00 << SDA | 0x01 << SCL,
+                    0x00 << SDA | 0x00 << SCL,
+                ],
+                Transfer::Stop => vec![
+                    0x00 << SDA | 0x01 << SCL,
+                    0x01 << SDA | 0x00 << SCL,
+                    0x01 << SDA | 0x01 << SCL,
+                ],
+                Transfer::Addr(byte, read) => self.byte::<SDA, SCL>((byte << 1) | read as u8, read),
+                Transfer::Write(bytes) => bytes
+                    .iter()
+                    .flat_map(|byte| self.byte::<SDA, SCL>(*byte, false))
```

Comment:
I think I am lost here. Is this function here to simulate I2C traffic or generate bits to send (and thus have to keep SDA high for the device to pull it low)?

Reply:
I think that you commented on the Read before I push some updates, right?
The Read should only generate the clock and keep the sda high so that the target can pull it down.